iwyu.cc

```diff
@@ -2578,6 +2578,12 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
   // Visitors defined by BaseAstVisitor.
 
   bool VisitNestedNameSpecifier(NestedNameSpecifier* nns) {
+    // If this is a NamespaceAlias, then mark the alias as used,
```

Comment:
Hmm, out of curiosity, why isn't the check for `NestedNameSpecifier::Namespace` necessary here too?

Reply:
I'm afraid you've lost me. This hunk is about detecting when an identifier is preceded by an explicit namespace alias, in which case we want to report the use of the alias.

As far as I understand it, we can't detect using-directive usage here (at least not in a way that tracks usage completely).

Reply:
Right, so:
```
using alias = std;

void foo() {
    alias::vector<int> x;  // hits your VisitNestedNameSpecifier check
    std::vector<int> x;  // why don't we want to check this in VisitNestedNameSpecifier too?
}
```

Reply:
Note that this hunk will be removed once we drop commit 2 i.e. warning on unused aliases.

In VisitNestedNameSpecifier() we're able to identify alias use (and therefore spot unused aliases).

We could consider std::vector too, and try to identify potential using-directives that are being used. But this won't be able to catch all uses appropriately. std::vector specifically is generally not 'using' any using-directive. As I think I've mentioned previously, identifying where a using-directive is used requires hooking into the symbol lookup infrastructure of clang (Sema/SemaLookup.cpp).

Reply:
> Note that this hunk will be removed once we drop commit 2

I got confused between the commits - this was in the first commit so still being added.

Reply:
Yeah, I'm not saying we should do anything, I was just curious why alias uses were different from other nested name-specifiers. Maybe because a namespace nns ultimately refers to a symbol that is guaranteed to live in the same file as the namespace. The alias, on the other hand, could have been included from somewhere that has nothing to do with the aliased namespace/symbol. OK, that makes sense to me, whether it's correct or not :)